### PR TITLE
create samples with padding to avoid truncations

### LINKED ISSUE #192: Option to avoid truncations while packing
# 🎯 **Goal (What & Why)**
Training on truncated documents can adversely affect downstream task performance, especially during instruction finetuning. 

# 🚀 **Execution Plan**
Add an option which controls whether documents will be truncated or not. In case of no truncations, if the document does not fit in the current sequence, it will be put in a new sequence. Documents longer than sequence length will be skipped.

# 📌 **Acceptance Criteria** (Must-Haves for Completion)
* The feature must be **functional and tested**.  

# 🛠️ **Project Management**
- [ ] **Assign the project to the Fast-LLM project.**
- [ ] **Set the `Estimate` field (in days) in the GitHub project.**
- [ ] **Use the `Size` field to categorize the PR size (Small/Medium/Large).**
- [ ] **Assign an owner when opening the issue.**

### PR DESCRIPTION
**Pull request by sohamparikh**
# ✨ Description

Provide an option to prevent document truncations while packing into sequences. Preventing truncations has shown to significantly boost downstream model quality, especially in supervised fine-tuning. Turning this on will hurt the training throughput.

Closes #192 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

### Testing

- [x] 🧪 I have added or updated tests to cover my changes.
- [x] ✔️ New and existing tests pass locally with my changes.
- [x] 🚦 I have tested these changes on GPUs and verified training stability.
- [ ] 🏋️ I have tested the changes on realistic training workloads, if applicable.

### Performance Impact

- [x] 📊 I have run benchmarks where applicable to evaluate the performance impact.
- [ ] ✅ The benchmarks show no performance regression.
- [ ] 🚀 The benchmarks indicate a potential performance improvement.
- [ ] ⚠️ The benchmarks indicate a potential performance degradation.
- [ ] 📈 I have provided benchmark results and detailed any performance impact below, if applicable.

### CODE REVIEW

File: fast_llm/csrc/data.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@
 
 /*
  Helper methods for fast index mapping builds.
- Changes for Fast-LLM: Use int16 for dataset index, add verbose argument to build_sample_idx.
+ Changes for Fast-LLM: Use int16 for dataset index, add verbose argument to build_sample_idx, add build_sample_idx_padded
```
Comment (jlamypoirier) [inline fast_llm/csrc/data.cpp:30]: I was hoping to get rid of this file... But I guess it's more difficult for `build_sample_idx_padded`

File: fast_llm/data/dataset/gpt/sampled.py [hunk 2]
```diff
@@ -362,6 +362,183 @@ def _load_yaml_data(self, data: dict[str, typing.Any]):
         self._unshuffled_documents = data["unshuffled_epochs"] * self._documents_per_epoch
 
 
+class GPTSampledIndexedPaddedDataset(SampledDataset):
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:365]: There is a lot of code in common with `GPTSampledIndexedDataset`. Could we merge the two classes, or at least create a common base class to reduce duplication?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 3]
```diff
@@ -362,6 +362,183 @@ def _load_yaml_data(self, data: dict[str, typing.Any]):
...
+        elif self._config.shuffle == ShufflingType.disabled:
+            doc_idx = np.tile(filtered_epoch_idx, num_epochs)
+        else:
+            raise NotImplementedError(f"Unknown shuffling type: {self._config.shuffle}")
+
+        assert _extension_available, (
+            "The C++ extension for dataset sampling is missing." " Please make sure Fast-LLM is installed correctly."
+        )
+
+        sample_idx = build_sample_idx_padded(
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:477]: Would it be possible to build a padded token cumsum instead? (python or cpp). This would make things almost identical to `GPTSampledIndexedDataset`

File: fast_llm/csrc/data.cpp [hunk 4]
```diff
@@ -129,6 +129,74 @@ py::array build_sample_idx(const py::array_t<int32_t>& sizes_,
...
+        cout << "     total number of samples:   " << num_samples << endl << std::flush;
+    }
+
+    int64_t sample_index = 0;
+    int64_t doc_idx_index = 0;
+    int64_t start_doc_index = 0;
+    int32_t n_docs = 0;
+    while (sample_index <= num_samples) {
+      int32_t remaining_seq_length = seq_length + 1;
+      while (remaining_seq_length != 0) {
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/csrc/data.cpp:166]: I wrote this with the assumption that all documents are `seq_length + 1` or shorter (added in the docstring as well). That makes it easier to use the returned `sample_idx` in a continuous fashion, otherwise there could be discontinuities and the caller of this function would also need to check for inclusion of a document (or we will need to return all indices in a sequence which is wasteful)

I can throw an error or warning if we see documents longer than `seq_length + 1`

File: fast_llm/csrc/data.cpp [hunk 5]
```diff
@@ -129,6 +129,74 @@ py::array build_sample_idx(const py::array_t<int32_t>& sizes_,
 
 }
 
+py::array build_sample_idx_padded(const py::array_t<int32_t>& sizes_,
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/csrc/data.cpp:132]: The returned values are semantically different so I kept them separate for readability

File: fast_llm/data/dataset/gpt/sampled.py (hunk 3 above)
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py:477]: I wanted to do that initially, but couldn't think of an efficient way to get the number of padded tokens for each sequence without explicitly looping. I can instead create an issue for it?

Comment (jlamypoirier) [reply fast_llm/data/dataset/gpt/sampled.py:502]: I meant with looping, I don't think there is any way around it.

File: fast_llm/csrc/data.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/csrc/data.cpp:166]: This is bound to happen, so we can't just have an infinite loop, and a crash would be really annoying. So we can either crop or ignore such sequences, maybe with a warning. Either case should be simple to handle in `__getitem__`-

File: fast_llm/data/dataset/gpt/sampled.py [hunk 6]
```diff
@@ -73,10 +74,16 @@ def _lazy_load(self):
 TOKEN_CUMSUM_RATE = 10
 
 
-class GPTSampledIndexedDataset(SampledDataset):
-    """
-    A sampled GPT dataset.
-    """
+class GPTSampledDatasetFactory:
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:77]: Unnecessary wrapper?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 7]
```diff
@@ -92,32 +99,93 @@ def __init__(
...
-            self._yaml_path = base_path.with_suffix(".yaml")
+            self._load_cache(base_path)
             # Sample or validate the dataset of a given rank.
             if sampling.distributed.config.rank == sampling.get_next_rank():
                 self._sample()
             # No barrier yet to allow running in parallel.
             # There needs to be one before calling `__getitem__`, normally handled through `GPTData`.
 
+    @abstractmethod
+    def _setup_cache(self):
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:116]: Please add return type hints, they are especially important for abstract methods

File: fast_llm/data/dataset/gpt/sampled.py [hunk 8]
```diff
@@ -419,6 +627,7 @@ def _sample(self) -> None:
         """
         logger.info(f" > Sampling dataset {self._indexed_dataset.name} ...")
         document_sizes = self._indexed_dataset.get_document_sizes()
+        doc_idx = np.arange(document_sizes.size, dtype=np.int32)
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:630]: Unncessary diff

File: fast_llm/data/dataset/gpt/sampled.py (hunk 3 above)
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py]: I've merged padding into `GPTSampledIndexedDataset` and it uses token_cumsum with some additional artifacts to track the number of padded samples. Have to get cumsum in cpp, python is too slow. 


We can move padding to a child class of `GPTSampledIndexedDataset` (or bring back the base class) if you think it's too complicated...

File: fast_llm/csrc/data.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (tscholak) [reply fast_llm/csrc/data.cpp]: in the use case for this we will never have to deal with docs longer than seq_length, because these will be thrown away.

File: fast_llm/data/dataset/gpt/sampled.py [hunk 9]
```diff
@@ -356,11 +433,31 @@ def _lazy_load(self):
         if not hasattr(self, "_documents_per_epoch"):
             self._load_yaml_data(yaml.safe_load(self._yaml_path.open("r")))
 
-    def _load_yaml_data(self, data: dict[str, typing.Any]):
+    def _load_yaml_data(self, data: dict[str, typing.Any]) -> None:
         self._documents_per_epoch = data["dataset"]["documents_per_epoch"]
         self._unshuffled_tokens = data["unshuffled_epochs"] * data["dataset"]["tokens_per_epoch"]
         self._unshuffled_documents = data["unshuffled_epochs"] * self._documents_per_epoch
 
+    def _load_verify_yaml_data(self, data: dict[str, typing.Any]) -> None:
```
Comment (tscholak) [inline fast_llm/data/dataset/gpt/sampled.py:441]: that's a nice addition!

File: fast_llm/data/dataset/gpt/sampled.py [hunk 10]
```diff
@@ -312,25 +360,52 @@ def __getitem__(self, index: int) -> typing.Any:
...
+                    )
+                    token_ids.append(sample.token_ids)
+                    if self._config.use_loss_masking_spans:
+                        for loss_masking_span in sample.loss_masking_spans:
+                            span = np.clip(loss_masking_span + token_count - token_start, 0, self._sequence_length + 1)
+                            if span[1] > span[0]:
+                                loss_masking_spans.append(span)
+                else:
+                    if document_size > self._sequence_length + 1:
+                        log_main_rank(
```
Comment (tscholak) [inline fast_llm/data/dataset/gpt/sampled.py:385]: this may happen so often that this will flood logs

File: fast_llm/data/dataset/gpt/sampled.py [hunk 11]
```diff
@@ -312,25 +360,52 @@ def __getitem__(self, index: int) -> typing.Any:
...
+                        token_ids.append(np.full((token_end - token_count,), -100, dtype=np.int64))
+                        token_count = token_end
+                    else:
+                        sample = self._indexed_dataset.get(
+                            document_index,
+                            use_loss_masking_spans=self._config.use_loss_masking_spans,
+                        )
+                        token_ids.append(sample.token_ids)
+                        if self._config.use_loss_masking_spans:
+                            loss_masking_spans.extend(sample.loss_masking_spans)
```
Comment (tscholak) [inline fast_llm/data/dataset/gpt/sampled.py:403]: do the loss masking spans need to be shifted?

Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py:403]: yes, thank you!

**Review by jlamypoirier — CHANGES_REQUESTED** [review]
@sohamparikh Things were getting way too complicated, I made a few changes to make my previous suggestions more concrete. These should make things way simpler overall. I haven't tested it though, see also  comments in code

**Review by sohamparikh — COMMENTED** [review]
Thanks @jlamypoirier, this is much simpler! I have a few fixes in the suggestions below, let me know if it makes sense. 
Mainly using `sequence_length + 1` everywhere for padding

File: fast_llm/csrc/data.cpp [hunk 12]
```diff
@@ -129,6 +129,64 @@ py::array build_sample_idx(const py::array_t<int32_t>& sizes_,
...
+  while (sizes_idx < sizes.size()) {
+    int32_t size = sizes[sizes_idx];
+    if (size > seq_length+1) {
+      // Skip sequences that are too long, to avoid truncations
+      sizes_idx += 1;
+    } else if (seq_size + size > seq_length) {
+      // add padded tokens if a document does not fit in current sequence and start a new sequence
+      cumsum += seq_length - seq_size;
+      seq_size = 0;
+    } else {
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/csrc/data.cpp:158]: Not in the cpp file, I made it pass `seq_length + 1` as argument instead.

File: fast_llm/data/dataset/gpt/sampled.py [hunk 13]
```diff
@@ -231,51 +249,72 @@ def _sample(self) -> None:
...
+            # Calculate the cumsum.
+            out.cumsum_(0)
+            # Crop unnecessary entries.
+            out = out[
+                : torch.clamp_min_(
+                    torch.searchsorted(out, self._num_samples * self._sequence_length, side="right"),
+                    0,
+                )
+            ]
+            return out.numpy(force=self._config.gpu), None
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/data/dataset/gpt/sampled.py:307]: It's useless at the end of a function, python deletes it automatically.

File: fast_llm/data/dataset/gpt/sampled.py [hunk 14]
```diff
@@ -287,6 +326,9 @@ def __getitem__(self, index: int) -> typing.Any:
         The returned sample is ready to be concatenated, then fed to a `GPTModel` (see `GPTModel.preprocess`).
         """
         self._lazy_load()
+        # tokens at the boundary are included in only one sample when we pack without truncations
+        # in case of packing with truncations, the last token from the previous sample is also the first token of the next sample
+        # TODO!!!!!!!!! Wrong for `not self._truncate_documents` (self._sequence_length + 1)
         token_start = index * self._sequence_length
```
[the remark this answers is not in the document]
Comment (jlamypoirier) [reply fast_llm/data/dataset/gpt/sampled.py:332]: Yes, left it unchanged because t needs to be fixed in several other places (like next line), can you please review this?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 15]
```diff
@@ -126,6 +127,22 @@ def _sample(self) -> None:
...
 
         # Calculate basic stats.
+        if not self._truncate_documents:
+            assert _extension_available, (
+                "The C++ extension for dataset sampling is missing."
+                " Please make sure Fast-LLM is installed correctly."
+            )
+            # NOTE!!!!!!!!! This won't work, it will mess up document indices. Also not really needed.
+            # length_filter = document_sizes <= self._sequence_length
+            # document_sizes = document_sizes[length_filter]
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py:137]: Got it, I did have that mapping as `document_indices`

File: fast_llm/data/dataset/gpt/sampled.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py]: Looks like this is the only place. Will post a confirmation once I test on a realistic workload

**Comment by sohamparikh** [comment]
LGTM now. 

Verified that the sampling is as expected with tons of simulation and simple python packing logic. 
The loss is also as expected and in the same range as we get from Llama-2-Accessory

image

**Comment by tscholak** [comment]
Awesome! Could we use these tests for unit tests?

**Comment by sohamparikh** [comment]
@jlamypoirier i believe we would need to filter long documents for correct sampling (and bring back the index map). In cases with too many long documents, or when they're highly concentrated around some indices document_sampling_index `token_start_cumsum_index * TOKEN_CUMSUM_RATE` can be low enough to re-sample some documents.

For example, with `seqlen=4` and `doc_sizes=[8, 2, 9, 1, 6, 7, 4, 1, 2, 9, 10, 11, 4, 9, 9, ...]` and `token_cumsum_rate=3`, we should have samples `{0: [2, 1], 1: [4,1], 2: [2], ...}`
Since `token_cumsum=[0, 9, 20,...]` here, at `index=2` we have `token_count=9`, `token_start=10` and `document_sampling_index=3` which will pick `[4,1]` instead of `[2]`

Found these issues when I also included longer documents in my tests.

**Comment by jlamypoirier** [comment]
> @jlamypoirier i believe we would need to filter long documents for correct sampling (and bring back the index map). In cases with too many long documents, or when they're highly concentrated around some indices `token_start_cumsum_index * TOKEN_CUMSUM_RATE` can be low enough to re-sample some documents.

That seems to be an error in `build_padded_token_cumsum`. It's excluding the long documents from the sample count but should include them (with length 0) so the index matches.

**Comment by sohamparikh** [comment]
> it would be even better in a test suite
Do you mean adding a unit test with the example I shared earlier, or something else?

The tests I've added already cover these kind of cases, I just presented a simpler example

**Comment by tscholak** [comment]
You were saying earlier that you're doing some end-to-end testing with simulated sequences and a separate greedy implementation. I'd like to see that as a test in Fast-LLM.

**Comment by sohamparikh** [comment]
`test_gpt_sample_padding` is exactly that. 
We can simulate many more seeds (~1000) as well, but it might add take a minute or two to run. If that's okay, I can do that

**Comment by jlamypoirier** [comment]
I don't think that's a seed issue. The bug should have happened whenever there are discarded documents, and that clearly happens in the test, so something else must be missing...

**Comment by tscholak** [comment]
> `test_gpt_sample_padding` is exactly that. We can simulate many more seeds (~1000) as well, but it might add take a minute or two to run. If that's okay, I can do that

We can have slow tests. We just need to mark them as such so that they become optional.

**Comment by sohamparikh** [comment]
> I don't think that's a seed issue. The bug should have happened whenever there are discarded documents, and that clearly happens in the test, so something else must be missing...


Adding more seeds can just cover more edge cases that aren't present in the few seeds. The latest bug didn't occur in the few initial seeds I'd tried, even though there were long documents

**Comment by jlamypoirier** [comment]
The bug discussed above is a common case, not an edge case. If the tests didn't fail before it means something is wrong with the test and adding more seeds won't change anything.

We can have slow tests (marked as such), but only if really needed as they can slow down development. (We need to make sure all tests pass or each PR)

File: tests/data/test_sampling.py [hunk 16]
```diff
@@ -126,3 +134,57 @@ def test_gpt_sample(seed, shuffle):
...
+    token_cumsum = build_padded_token_cumsum(sizes, sequence_length + 1, token_cumsum_rate, offset)
+    Assert.all_equal(token_cumsum, expected_cumsums)
+
+
+@pytest.mark.skipif(not _extension_available, reason="CPP Extension not available")
+@pytest.mark.parametrize("seed", (0, 1234, 5678, 571, 687, 1, 188))
+def test_gpt_sample_padding(seed):
+    vocab_size = 131072
+    np.random.seed(seed)
+    num_sequences = np.random.randint(1, 1000)
```
Comment (jlamypoirier) [inline tests/data/test_sampling.py:159]: No need for such big numbers. O(10) seqlens and sequence counts should be enough.

Comment (sohamparikh) [reply tests/data/test_sampling.py:159]: we wont need to mark it as slow in that case...

**Comment by sohamparikh** [comment]
@jlamypoirier do the changes look ok now? 

I don't see anything wrong with the python verification. To be clear about the bugs caught with it, the long documents issue wasn't caught in the first 1-2 seeds I tried, but saw it in after I added a few more manually. The larger simulations caught the bug where `num_tokens_unshuffled` for padding was computed the same way as before

File: fast_llm/data/dataset/gpt/sampled.py [hunk 17]
```diff
@@ -231,51 +245,78 @@ def _sample(self) -> None:
...
+                )
+            ]
+            return out.numpy(force=self._config.gpu), None
+        else:
+            # TODO: dynamically handle int64 or int32 in CPP
+            out = build_padded_token_cumsum(
+                sizes.cpu().numpy(), (self._sequence_length + 1), TOKEN_CUMSUM_RATE, offset
             )
-        ]
+            num_tokens = out[-1]
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:315]: Since we are throwing away long documents, we could end up not generating enough tokens. At the very least we need to add a check for it. Or maybe exclude long documents from `tokens_per_epoch`?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 18]
```diff
@@ -379,6 +447,12 @@ def __init__(
         self._indexed_dataset = indexed_dataset
         self._num_samples = sampling.num_samples
         self._sequence_length = sampling.sequence_length
+        self._config = sampling.config
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:450]: Unused?

File: tests/data/test_sampling.py [hunk 19]
```diff
@@ -126,3 +134,63 @@ def test_gpt_sample(seed, shuffle):
...
+
+def get_test_seeds(num_seeds):
+    np.random.seed(42)
+    seeds = np.random.randint(0, num_seeds * 100, num_seeds)
+    return seeds.tolist()
+
+
+@pytest.mark.skipif(not _extension_available, reason="CPP Extension not available")
+def test_gpt_sample_padding():
+    for seed in get_test_seeds(10000):
```
Comment (jlamypoirier) [inline tests/data/test_sampling.py:162]: 100 seeds should be more than enough. Anything more than that is just a waste of compute.

File: fast_llm/data/dataset/gpt/sampled.py [hunk 20]
```diff
@@ -231,51 +245,84 @@ def _sample(self) -> None:
...
+            return out.numpy(force=self._config.gpu), None
+        else:
+            # TODO: dynamically handle int64 or int32 in CPP
+            out = build_padded_token_cumsum(
+                sizes.cpu().numpy(), (self._sequence_length + 1), TOKEN_CUMSUM_RATE, offset
             )
-        ]
+            num_tokens = out[-1]
+            # TODO: should this instead be a low multiple of seqlen + 1, or a fraction of num_samples?
+            if num_tokens <= self._sequence_length + 1:
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/sampled.py:317]: Doesn't make sense. We need the whole `self._num_samples * (self._sequence_length + 1)` to be generated, otherwise training will fail. But the check can't be here because what we care about is the sum of shuffled and unshuffled tokens.

Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py:317]: ah yes, makes sense

Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py]: It made more sense to compute `tokens_per_epoch` like you said. Since it's a lower bound on the actual number of tokens (including padding), `num_epochs` is now an upper bound on the actual number of epochs needed and I don't think we need to catch it anymore. This could end up being wasteful in case of a lot of padding, but would avoid failure during training

**Review by jlamypoirier — APPROVED** [review]
